### PR TITLE
Modify getInitialData method to fetch extra data required to render a…

### LINKED ISSUE #6252: Data fetching for secondary column onward journeys
**Is your feature request related to a problem? Please describe.**
At present the Features and Analysis and Top Stories components that sit within the secondary column on story pages are populated using fixture data, e.g.: https://github.com/bbc/simorgh/blob/74221208cd41fc4f6c63aa1ecbec0b75a2794f92/src/app/containers/CpsFeaturesAnalysis/index.jsx#L5

Soon a new Ares endpoint will be available that will provide us with the required live data.

While we await the new endpoint, we can begin work on designing the actual mechanism for fetching this data and passing it to the components that need to use it.

**Describe the solution you'd like**
- Create a local test api serving the fixture data specified above, e.g https://www.npmjs.com/package/http-server 
- Look at existing patterns for fetching ancillary page data (e.g. Includes, Most Read, the radio schedule component)
  - We don't need to request the data on a per asset basis, the data is common to all Story pages per service
  - Request must be serverside, ruling out clientside solutions such as `useEffect`
  - Consider how a clientside fetch would work, when clientside links are enabled
- Confer with the Home team about their approaches to ensure consistency
- Recommend a solution and create a proof of concept PR
- Ensure that the solution will scale to support the different components that need this data

**Describe alternatives you've considered**

**Testing notes**
No formal testing for POC

Dev insight: Will Cypress tests be required or are unit tests sufficient? Will there be any potential regression? etc

- [x] This feature is expected to need manual testing.

**Additional context**

### PR DESCRIPTION
**Pull request by chris-hinds**
… STY asset

Resolves https://github.com/bbc/simorgh/issues/6252

**Overall change:**
Introduces a new method to getInitialData for the CPSAsset route, allowing us to declare multiple endpoints that are required to construct a given asset type ensuring all necessary data is fetched on the server side render.

This PR enables the fetching of mostRead data along side a CPS STY asset. Soon we will also add the secondary column onward journey endpoint.

**Code changes:**

- Add util function to get asset type
- Add function to fetch multiple URL's and key the data with a name i.e. mostRead
- Merge additional data into existing pageData object
- Replaces the use of mocking the global fetch object where required, as some page types now require more than one endpoint to be fetched and mocking the global fetch object does not cater for this. fetchMock is already in use in other ares of the application.

---

- [X] I have assigned myself to this PR and the corresponding issues
- [X] I have added labels to this PR for the relevant pod(s) affected by these changes
- [X] I have assigned this PR to the Simorgh project

**Testing:**

- [ ] Automated (jest and/or cypress) tests added (for new features) or updated (for existing features)
- [ ] If necessary, I have run the local E2E non-smoke tests relevant to my changes (`CYPRESS_APP_ENV=local CYPRESS_SMOKE=false npm run test:e2e:interactive`)
- [ ] This PR requires manual testing

### CODE REVIEW

File: src/app/routes/cpsAsset/utils/getAdditionalPageData.js [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+          path: `/${service}/mostread`,
+        },
+      ];
+    default:
+      return null;
+  }
+};
+
+const validateResponse = ({ status, json }, name) => {
+  if (status == 200 && !isEmpty(json)) {
```
Comment (andrewscfc) [inline src/app/routes/cpsAsset/utils/getAdditionalPageData.js]:
```suggestion
  if (status === 200 && !isEmpty(json)) {
```
Is the linting allowing a `==` here?

File: src/app/routes/cpsAsset/utils/getAdditionalPageData.js [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    return { [name]: json };
+  }
+
+  return null;
+};
+
+const fetchUrl = ({ name, path }) =>
+  fetchPageData(path).then(response => validateResponse(response, name));
+
+const getAdditionalPageData = async (pageData, service, variant) => {
```
Comment (andrewscfc) [inline src/app/routes/cpsAsset/utils/getAdditionalPageData.js:32]: Should we include variant at this point as it's not currently used? I get that it is illustrative as to what we will need in the future.

File: src/app/routes/cpsAsset/utils/getAdditionalPageData.test.js [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+import mapJson from '#data/pidgin/cpsAssets/media-23256549.json';
+import styJson from '#data/mundo/cpsAssets/23263889.json';
+import mostReadJson from '#data/mundo/mostRead/index.json';
+
+describe('getAdditionalPageData', () => {
+  afterEach(() => {
+    jest.clearAllMocks();
+  });
+
+  it('should return no aditional data for a MAP asset', async () => {
```
Comment (andrewscfc) [inline src/app/routes/cpsAsset/utils/getAdditionalPageData.test.js]:
```suggestion
  it('should return no additional data for a MAP asset', async () => {
```

File: src/app/routes/cpsAsset/utils/getAdditionalPageData.test.js [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+    const expectedOutput = {
+      mostRead: mostReadJson,
+    };
+
+    expect(additionalPageData).toEqual(expectedOutput);
+  });
+
+  it('should return an empty object when a data fetch fails', async () => {
+    jest.spyOn(global, 'fetch').mockResponse(JSON.stringify({}));
```
Comment (andrewscfc) [inline src/app/routes/cpsAsset/utils/getAdditionalPageData.test.js]: Is that a non-200 response? I'd expect a failure to have a different error code?

**Review by andrewscfc — COMMENTED** [review]
Looks a good approach, could/should we apply this to the front page's radio schedules?

File: src/app/routes/cpsAsset/utils/getAdditionalPageData.js (hunk 1 above)
Comment (chris-hinds) [reply src/app/routes/cpsAsset/utils/getAdditionalPageData.js]: Not sure how that slipped through linting :/

File: src/app/routes/cpsAsset/utils/getAdditionalPageData.test.js (hunk 4 above)
Comment (chris-hinds) [reply src/app/routes/cpsAsset/utils/getAdditionalPageData.test.js]: still working on this test :)

File: src/app/routes/cpsAsset/utils/getAdditionalPageData.js [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+          path: `/${service}/mostread`,
+        },
+      ];
+    default:
+      return null;
+  }
+};
+
+const validateResponse = ({ status, json }, name) => {
+  if (status === 200 && !isEmpty(json)) {
```
Comment (andrewscfc) [inline src/app/routes/cpsAsset/utils/getAdditionalPageData.js:22]: Just thinking about status codes will the most read endpoint always return 200 in the success case? One case I can think of is a 202 response on the first request for most-read for a service; will the first cache of a page have a missing 'most read' section? Once the page cache expires this situation would be resolved as the next call to most-read would issue a 200 with the content. This endpoint may not 202, just asking if we know this and if so, are we okay with the above?

File: src/app/routes/index.test.jsx [hunk 6]
```diff
@@ -14,8 +21,14 @@ import legacyMediaAssetPage from '#data/azeri/legacyAssets/multimedia/2012/09/12
 import photoGalleryPageJson from '#data/indonesia/cpsAssets/indonesia-41635759.json';
 import storyPageJson from '#data/mundo/cpsAssets/noticias-internacional-51266689.json';
 import featureIndexPageJson from '#data/afrique/cpsAssets/48465371.json';
+import storyPageMostReadData from '#data/pidgin/mostRead/index.json';
 
-afterEach(() => jest.clearAllMocks());
+fetchMock.config.fallbackToNetwork = true; // ensures non mocked requests fallback to an actual network request
```
Comment (andrewscfc) [inline src/app/routes/index.test.jsx:26]: I've never been a fan of relying live or test endpoint in unit tests; it means that if an envinronment is down, your unit tests won't pass, giving a false indication that the code in this unit has broken. Additionally, if you are working offline, your unit tests will fail.

I can't actually see any code that needs access to the network, am I missing something?

**Review by andrewscfc — COMMENTED** [review]
Looks great, just a status code 'edge case' question and test query 🙂

File: src/app/routes/index.test.jsx (hunk 6 above)
Comment (chris-hinds) [reply src/app/routes/index.test.jsx:26]: Previously we mocked the global fetch object which meant that any fetch which occurred always returned the one mock response we had per test (until we reset the mock). Switching to fetchMock allows us to mock multiple fetches based on their url, this showed that our pages make a few different network requests, posts to analytics providers etc and we were only ever mocking the main pagedata.

I would like one to ensure we mock all of our requests however I think that would be quite a large refactor and not in scope for this pr.

File: src/app/routes/cpsAsset/utils/getAdditionalPageData.js (hunk 5 above)
Comment (chris-hinds) [reply src/app/routes/cpsAsset/utils/getAdditionalPageData.js:22]: not sure on the status codes actually, will check with the ares team. however if the first request was to return a non 200 we will simply not append the newly fetched data to the page data, and in the case of mostRead it will fallback to its client side data fetch using `useEffect()`

Comment (andrewscfc) [reply src/app/routes/cpsAsset/utils/getAdditionalPageData.js:22]: Hmm, if we fallback to `useEffect` I think we'd experience a big traffic spike to the endpoint to the most read endpoint until the initial page cache expires? I imagine that's okay, but just making sure we've thought this through.

File: src/app/routes/index.test.jsx (hunk 6 above)
Comment (andrewscfc) [reply src/app/routes/index.test.jsx:26]: Oh so this is no 'worse' than the position we are in now on `latest`? Happy to resolve this if that is the case, just wanted to avoid introducing a new dependency on the network if one was being introduced.

File: src/app/routes/cpsAsset/getInitialData/index.js [hunk 7]
```diff
@@ -52,13 +54,19 @@ const transformJson = async json => {
...
 };
 
-export default async ({ path: pathname }) => {
+export default async ({ path: pathname, service, variant }) => {
   const { json, ...rest } = await fetchPageData(pathname);
 
+  const additionalPageData = await getAdditionalPageData(
+    json,
+    service,
+    variant,
```
Comment (mykeels) [inline src/app/routes/cpsAsset/getInitialData/index.js:63]: Here, we're passing the `variant`, which does not exist as an argument in `getAdditionalPageData`.

See https://github.com/bbc/simorgh/pull/6338/files#diff-e0de94df014039e48a3a35d353d4336fR31

File: src/app/routes/cpsAsset/utils/getAdditionalPageData.js [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+    return { [name]: json };
+  }
+
+  return null;
+};
+
+const fetchUrl = ({ name, path }) =>
+  fetchPageData(path).then(response => validateResponse(response, name));
+
+const getAdditionalPageData = async (pageData, service) => {
```
Comment (mykeels) [inline src/app/routes/cpsAsset/utils/getAdditionalPageData.js]: I'm thinking the `variant` should be a parameter in this function

File: src/app/routes/cpsAsset/utils/getAdditionalPageData.js (hunk 5 above)
Comment (chris-hinds) [reply src/app/routes/cpsAsset/utils/getAdditionalPageData.js:22]: Thats how mostRead has been built. MostRead is current client side only.

File: src/app/routes/cpsAsset/utils/getAdditionalPageData.js (hunk 8 above)
Comment (chris-hinds) [reply src/app/routes/cpsAsset/utils/getAdditionalPageData.js]: I need to remove the variant param, as we are not using it yet. I will write a new issue to ensure we construct urls for variant based services too

File: src/app/routes/cpsAsset/utils/getAdditionalPageData.js [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+import getAssetType from './getAssetType';
+import fetchPageData from '../../utils/fetchPageData';
+
+const pageTypeUrls = (assetType, service) => {
+  switch (assetType) {
+    case STORY_PAGE:
+      return [
+        {
+          name: 'mostRead',
+          path: `/${service}/mostread`,
```
Comment (thekp) [inline src/app/routes/cpsAsset/utils/getAdditionalPageData.js]: Just a note for this path. I have created this Utils: https://github.com/bbc/simorgh/blob/52cbca753e7ca408fe169ea5ce970997f58b0e22/src/app/lib/utilities/getMostReadUrls/index.js#L1-L5 for the variant path

File: src/app/routes/index.test.jsx (hunk 6 above)
Comment (chris-hinds) [reply src/app/routes/index.test.jsx:26]: > Oh so this is no 'worse' than the position we are in now on `latest`? Happy to resolve this if that is the case, just wanted to avoid introducing a new dependency on the network if one was being introduced.

No `worse` at all. This just highlights how many requests we are not yet mocking.

File: src/app/routes/cpsAsset/utils/getAdditionalPageData.js (hunk 8 above)
Comment (chris-hinds) [reply src/app/routes/cpsAsset/utils/getAdditionalPageData.js]: have update to remove the variant param for now

File: src/app/routes/cpsAsset/utils/getAdditionalPageData.js (hunk 9 above)
Comment (chris-hinds) [reply src/app/routes/cpsAsset/utils/getAdditionalPageData.js]: ohhh thats nice, did not know. I will update this PR to use that util. Thanks @thekp

File: src/app/routes/cpsAsset/utils/getAdditionalPageData.js (hunk 8 above)
Comment (FK78) [reply src/app/routes/cpsAsset/utils/getAdditionalPageData.js]: @hindsc52 I believe you need to still remove variant from [here](https://github.com/bbc/simorgh/pull/6338/files/3046c6c3964309de9405a082d37f992e55f37adf#r419326543)

Comment (chris-hinds) [reply src/app/routes/cpsAsset/utils/getAdditionalPageData.js]: Good spot @FK78 although I have now added variant back in as I am using a pre-existing util that has the variant logic baked in. This saves a small future PR

**Comment by chris-hinds** [comment]
@andrewscfc @mykeels @FK78 @thekp I have updated this PR based on all of your comments :)

File: src/app/routes/cpsAsset/utils/getAdditionalPageData.js [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+import fetchPageData from '../../utils/fetchPageData';
+import { getMostReadEndpoint } from '#lib/utilities/getMostReadUrls';
+
+const pageTypeUrls = (assetType, service) => {
+  switch (assetType) {
+    case STORY_PAGE:
+      return [
+        {
+          name: 'mostRead',
+          path: getMostReadEndpoint({ service, variant }).replace('.json', ''),
```
Comment (mykeels) [inline src/app/routes/cpsAsset/utils/getAdditionalPageData.js:13]: I believe this is the util function you mentioned in a previous comment. Shouldn't the `variant` be passed into this function as an argument?

Comment (mykeels) [reply src/app/routes/cpsAsset/utils/getAdditionalPageData.js:13]: Previous comment: https://github.com/bbc/simorgh/pull/6338#discussion_r419523417

File: src/app/routes/cpsAsset/utils/getAssetType.test.js [hunk 11]
```diff
@@ -0,0 +1,25 @@
+import getAssetType from './getAssetType';
```
Comment (thekp) [inline src/app/routes/cpsAsset/utils/getAssetType.test.js:2]: I like the convention of separating our imports 👍 
```suggestion
// components to test
import getAssetType from './getAssetType';
```

**Review by thekp — APPROVED** [review]
LGTM, I like the usage of fetch-mock here. 
A good example for this issue: https://github.com/bbc/simorgh/issues/5960

File: src/app/routes/cpsAsset/utils/getAssetType.test.js (hunk 11 above)
Comment (chris-hinds) [reply src/app/routes/cpsAsset/utils/getAssetType.test.js:2]: :) Ive started doing, I find it hard to read through the imports when we sometimes have lists of like 20+

File: src/app/routes/cpsAsset/utils/getAdditionalPageData.js (hunk 10 above)
Comment (chris-hinds) [reply src/app/routes/cpsAsset/utils/getAdditionalPageData.js:13]: it is now @mykeels `getMostReadEndpoint` expects an object, service is required variant has an optional default.

**Review by andrewscfc — APPROVED** [review]

**Review by mykeels — APPROVED** [review]

**Review by FK78 — APPROVED** [review]

**Comment by paruchurisilpa** [comment]
Looks good to me..
